### PR TITLE
Fix/refactor unified target parent lookup

### PR DESCRIPTION
**Pull request by balazsbajorics**
This is a PR that changes no functionality but reorganizes the reparent-strategy-helpers code with the aim to make it more readable and more sustainable. 

This is part of us "wrapping up" the reorder/reparent, and we did not want to leave a mess behind.

I split the code to multiple files so the diff in theory should be fairly zero-based. I'll mark functions which I just verbatim copied, and I'll mark places that are new code (old behavior)

### CODE REVIEW

File: editor/src/components/canvas/canvas-strategies/strategies/reparent-strategy-helpers.ts [hunk 1]
```diff
@@ -102,10 +106,41 @@ export function reparentStrategyForParent(
...
 }
 
+function isAutoLayoutComaptibleWithSingleAxisReorder(
+  metadata: ElementInstanceMetadataMap,
+  parent: ElementPath,
+): boolean {
+  const newParentMetadata = MetadataUtils.findElementByElementPath(metadata, parent)
+  const parentIsFlexLayout = MetadataUtils.isFlexLayoutedContainer(newParentMetadata)
+  if (parentIsFlexLayout) {
+    return true
```
[the remark this answers is not in the document]
Comment (gbalint) [reply editor/src/components/canvas/canvas-strategies/strategies/reparent-strategy-helpers.ts:116]: Hmmm, I think I misunderstood this function's role, probably we need a better function name then

File: editor/src/components/canvas/canvas-strategies/strategies/reparent-as-static-strategy.tsx [hunk 2]
```diff
@@ -98,3 +116,140 @@ function getIdAndNameOfReparentToStaticStrategy(targetLayout: 'flex' | 'flow'):
       assertNever(targetLayout)
   }
 }
+
+export function applyStaticReparent(
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/reparent-as-static-strategy.tsx:120]: this function is just moved without changes from reparent-strategy-helpers.
this was the only place it was called from, so it was pointless for this to live in a separate helper

File: editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-property-changes.ts [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+import { ReparentStrategy } from './reparent-strategy-helpers'
+
+const propertiesToRemove: Array<PropertyPath> = [
+  PP.create(['style', 'left']),
+  PP.create(['style', 'top']),
+  PP.create(['style', 'right']),
+  PP.create(['style', 'bottom']),
+]
+
+export function getAbsoluteReparentPropertyChanges(
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-property-changes.ts:43]: these 3 functions are not modified, just moved to this dedicated file.

File: editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-strategy-helpers.ts [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+export type ReparentStrategy = 'REPARENT_AS_ABSOLUTE' | 'REPARENT_AS_STATIC'
+
+export type FindReparentStrategyResult = {
+  strategy: ReparentStrategy
+  isFallback: boolean
+  target: ReparentTarget
+}
+
+export function reparentStrategyForPaste(
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-strategy-helpers.ts:25]: this is the same as master, just with a new name to indicate that this is _only_ used by the paste action

File: editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-strategy-helpers.ts [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    }
+  } else {
+    return {
+      strategy: 'REPARENT_AS_ABSOLUTE',
+      isFallback: false,
+    }
+  }
+}
+
+export function findReparentStrategies(
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-strategy-helpers.ts:51]: same as master

File: editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-strategy-parent-lookup.ts [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+import { ReparentStrategy, ReparentSubjects, ReparentTarget } from './reparent-strategy-helpers'
+import { drawTargetRectanglesForChildrenOfElement } from './reparent-strategy-sibling-position-helpers'
+
+export type FindReparentStrategyResult = {
+  strategy: ReparentStrategy
+  isFallback: boolean
+  target: ReparentTarget
+}
+
+export function getReparentTargetUnified(
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-strategy-parent-lookup.ts:37]: this is the same behavior as the old massive spaghetti-function getReparentTargetUnified, but with the code split into sub-functions for better readability

File: editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-strategy-parent-lookup.ts [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+  const targetParentUnderPoint: ReparentTarget = findParentUnderPointByArea(
+    targetParentPath,
+    metadata,
+    canvasScale,
+    pointOnCanvas,
+  )
+  return targetParentUnderPoint
+}
+
+function findValidTargetsUnderPoint(
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-strategy-parent-lookup.ts:87]: probably the most interesting change of the PR, this function was an inline filter function that was incredibly dense and mixed a few concepts and abstraction levels

Comment (balazsbajorics) [reply editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-strategy-parent-lookup.ts:87]: for comparison, the original inline array filtering can be found here https://github.com/concrete-utopia/utopia/blob/0452ce132ca5a1524e8685403d6922297dcdff8d/editor/src/components/canvas/canvas-strategies/strategies/reparent-strategy-helpers.ts#L320-L369

File: editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-strategy-parent-lookup.ts [hunk 8]
```diff
@@ -0,0 +1,482 @@
...
+
+function findParentUnderPointByArea(
+  targetParentPath: ElementPath,
+  metadata: ElementInstanceMetadataMap,
+  canvasScale: number,
+  pointOnCanvas: CanvasPoint,
+) {
+  const autolayoutDirection = singleAxisAutoLayoutContainerDirections(targetParentPath, metadata)
+  const shouldReparentAsFlowOrStatic = autoLayoutParentAbsoluteOrStatic(metadata, targetParentPath)
+  const compatibleWith1DReorder = isSingleAxisAutoLayoutComaptibleWithReorder(
```
Comment (balazsbajorics) [inline editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-strategy-parent-lookup.ts:309]: not every single axis autolayout is compatible with 1d reorder. specifically, for Flow if there's only a single children, we can't use that to determine direction, and in our testing with @gbalint we realized that the behavior just becomes unintuitive overall. so we only allow 1d reorder-like insertion if there's >1 static children for flow parents. for flow parents that have 0 or 1 children, we still allow static reorder, but we don't run the insertion indicators and the padded area checks

Comment (balazsbajorics) [reply editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-strategy-parent-lookup.ts:309]: this is the same behavior as on master today, but on master this is totally implicit in https://github.com/concrete-utopia/utopia/blob/0452ce132ca5a1524e8685403d6922297dcdff8d/editor/src/components/canvas/canvas-strategies/strategies/reparent-strategy-helpers.ts#L1062-L1067

**Review by enidemi — APPROVED** [review]

File: editor/src/components/canvas/canvas-strategies/strategies/reparent-as-static-strategy.tsx (hunk 2 above)
Comment (Rheeseyb) [reply editor/src/components/canvas/canvas-strategies/strategies/reparent-as-static-strategy.tsx:120]: Should the `export` modifier be deleted?

Comment (balazsbajorics) [reply editor/src/components/canvas/canvas-strategies/strategies/reparent-as-static-strategy.tsx:120]: ah yes!

**Review by gbalint — APPROVED** [review]

**Review by seanparsons — APPROVED** [review]

**Pull request merged**